File: src/eventql/db/Table.cc
/**
 * Copyright (c) 2016 zScale Technology GmbH <legal@zscale.io>
 * Authors:
 *   - Paul Asmuth <paul@zscale.io>
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License ("the license") as
 * published by the Free Software Foundation, either version 3 of the License,
 * or any later version.
 *
 * In accordance with Section 7(e) of the license, the licensing of the Program
 * under the license does not imply a trademark license. Therefore any rights,
 * title and interest in our trademarks remain entirely with us.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the license for more details.
 *
 * You can be released from the requirements of the license by purchasing a
 * commercial license. Buying such a license is mandatory as soon as you develop
 * commercial activities involving this program without disclosing the source
 * code of your own applications
 */
#include <eventql/db/Table.h>
#include <eventql/db/TimeWindowPartitioner.h>
#include <eventql/db/FixedShardPartitioner.h>

#include "eventql/eventql.h"

namespace eventql {

Table::Table(
    const TableDefinition& config) :
    config_(config) {
  loadConfig();
}

String Table::name() const {
  std::unique_lock<std::mutex> lk(mutex_);
  return config_.table_name();
}

String Table::tsdbNamespace() const {
  std::unique_lock<std::mutex> lk(mutex_);
  return config_.customer();
}

Duration Table::partitionSize() const {
  std::unique_lock<std::mutex> lk(mutex_);
  return 4 * kMicrosPerHour;
}

size_t Table::sstableSize() const {
  std::unique_lock<std::mutex> lk(mutex_);
  return config_.config().sstable_size();
}

size_t Table::numShards() const {
  std::unique_lock<std::mutex> lk(mutex_);
  return config_.config().num_shards();
}

Duration Table::commitInterval() const {
  std::unique_lock<std::mutex> lk(mutex_);
  return config_.config().commit_interval();
}

RefPtr<msg::MessageSchema> Table::schema() const {
  std::unique_lock<std::mutex> lk(mutex_);
  return schema_;
}

TableDefinition Table::config() const {
  std::unique_lock<std::mutex> lk(mutex_);
  return config_;
}

TableStorage Table::storage() const {
  std::unique_lock<std::mutex> lk(mutex_);
  return config_.config().storage();
}

const String& Table::getPartitionKey() const {
  std::unique_lock<std::mutex> lk(mutex_);
  return config_.config().partition_key();
}

TablePartitionerType Table::partitionerType() const {
  std::unique_lock<std::mutex> lk(mutex_);
  return config_.config().partitioner();
}

RefPtr<TablePartitioner> Table::partitioner() const {
  std::unique_lock<std::mutex> lk(mutex_);
  return partitioner_;
}

Vector<String> Table::getPrimaryKey() const {
  std::unique_lock<std::mutex> lk(mutex_);
  return Vector<String>(
      config_.config().primary_key().begin(),
      config_.config().primary_key().end());
}

MetadataTransaction Table::getLastMetadataTransaction() const {
  std::unique_lock<std::mutex> lk(mutex_);

  return MetadataTransaction(
      SHA1Hash(
          config_.metadata_txnid().data(),
          config_.metadata_txnid().size()),
      config_.metadata_txnseq());
}

void Table::updateConfig(TableDefinition new_config) {
  std::unique_lock<std::mutex> lk(mutex_);
  config_ = new_config;
  loadConfig();
}

void Table::loadConfig() {
  schema_ = msg::MessageSchema::decode(config_.config().schema());

  switch (config_.config().partitioner()) {

    case TBL_PARTITION_TIMEWINDOW:
      if (config_.config().has_time_window_partitioner_config()) {
        partitioner_ = RefPtr<TablePartitioner>(
            new TimeWindowPartitioner(
                config_.table_name(),
                config_.config().partition_key(),
                config_.config().time_window_partitioner_config()));
      } else {
        partitioner_ = RefPtr<TablePartitioner>(
            new TimeWindowPartitioner(
                config_.table_name(),
                config_.config().partition_key()));
      }
      break;

    case TBL_PARTITION_FIXED:
      partitioner_ = RefPtr<TablePartitioner>(
          new FixedShardPartitioner(
              config_.table_name(),
              config_.config().num_shards()));
      break;

  }
}

}

File: src/eventql/webui/views/sql_editor/table.js
/**
 * Copyright (c) 2017 DeepCortex GmbH <legal@eventql.io>
 * Authors:
 *   - Laura Schlimmer <laura@eventql.io>
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License ("the license") as
 * published by the Free Software Foundation, either version 3 of the License,
 * or any later version.
 *
 * In accordance with Section 7(e) of the license, the licensing of the Program
 * under the license does not imply a trademark license. Therefore any rights,
 * title and interest in our trademarks remain entirely with us.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the license for more details.
 *
 * You can be released from the requirements of the license by purchasing a
 * commercial license. Buying such a license is mandatory as soon as you develop
 * commercial activities involving this program without disclosing the source
 * code of your own applications
 */

EventQL.SQLEditor.ResultList.TableChartBuilder = function(elem, params) {
  'use strict';

  var data_visualizations = [
    {
      key: "table",
      title: "Table"
    },
    {
      key: "line",
      title: "Linechart"
    },
    {
      key: "bar",
      title: "Barchart"
    }
  ];

  var on_csv_download = [];
  var on_param_change = [];

  this.onCSVDownload = function(fn) {
    on_csv_download.push(fn);
  };

  this.onParamChange = function(fn) {
    on_param_change.push(fn);
  };

  this.render = function(result) {
    var tpl = TemplateUtil.getTemplate(
        "evql-sql-editor-result-table-chart-builder-tpl");
    elem.appendChild(tpl);

    initToolbar();

    var data_visualization = getDataVisualizationValue();

    var dropdown_elem = elem.querySelector("[data-control='visualization']");
    var dropdown = new EventQL.Dropdown(dropdown_elem);
    dropdown.setMenuItems(data_visualizations);
    dropdown.setValue(data_visualization);

    dropdown.onChange(function(value) {
      updateDataVisualization(value, result);
      on_param_change.forEach(function(callback_fn) {
        callback_fn({visualization: value});
      });
    });

    updateDataVisualization(data_visualization, result);
  };

  function initToolbar() {
    elem.querySelector("[data-content='result_idx']").innerHTML = params.idx + 1;

    if (params.hidable) {
      var visibility_ctrl = elem.querySelector("[data-control='visibility']");
      visibility_ctrl.addEventListener("click", function(e) {
        elem.classList.toggle("hidden");
      }, false);
    }

    var download_csv_elem = elem.querySelector("[data-control='download_csv']");
    download_csv_elem.addEventListener("click", function(e) {
      on_csv_download.forEach(function(callback_fn) {
        callback_fn();
      });
    }, false);
  }

  function updateDataVisualization(visualization, result) {
    switch (visualization) {
      case "table":
        renderTable(result);
        break;

      case "bar":
      case "line":
        buildChart(result, visualization);
        break;
    }
  }

  function renderTable(result) {
    var thead_elem = document.createElement("tr");
    result.columns.forEach(function(col) {
      var th_elem = document.createElement("th");
      th_elem.innerHTML = DOMUtil.escapeHTML(col);
      thead_elem.appendChild(th_elem);
    });

    var tbody_elem = document.createElement("tbody");
    result.rows.forEach(function(row) {
      var tr_elem = document.createElement("tr");

      row.forEach(function(cell) {
        var td_elem = document.createElement("td");
        td_elem.innerHTML = DOMUtil.escapeHTML(cell);
        tr_elem.appendChild(td_elem);
      });

      tbody_elem.appendChild(tr_elem);
    });

    var table = document.createElement("table");
    table.classList.add("evql_table");
    var thead = document.createElement("thead");
    thead.appendChild(thead_elem);
    table.appendChild(thead);
    table.appendChild(tbody_elem);

    DOMUtil.replaceContent(elem.querySelector(".result_view"), table);
  }

  function buildChart(result, chart_type) {

    var chart_cfg = {
      height: 300
    };

    switch (chart_type) {
      case "bar":
        break;

      case "line":
        chart_cfg.timeseries = true; //FIXME make this configurable
        chart_cfg.points = true;
        break;

      case "area":
        chart_cfg.timeseries = true; //FIXME make this configurable
        break;

      default:
        console.log("ERROR unknown chart type", type);
        return;
    }

    chart_cfg.type = chart_type;

    var chart_elem = document.createElement("div")
    chart_elem.classList.add("chart_container");
    DOMUtil.replaceContent(elem.querySelector(".result_view"), chart_elem);

    var chart = new EventQL.ChartPlotter(chart_elem, chart_cfg);

    try {
      chart.render(result);

    /* render chart builder errror */
    } catch (e) {
      var error_elem = document.createElement("div");
      error_elem.classList.add("message", "error");
      error_elem.innerHTML = "The chosen chart can't be drawn. Please try another chart type or select table to display the result.";
      chart_elem.appendChild(error_elem);
    }
  }

  function getDataVisualizationValue() {
    var default_value = data_visualizations[0].key;

    if (params.data_visualization) {
      for (var i = 0; i < data_visualizations.length; i++) {
        if (data_visualizations[i].key == params.data_visualization) {
          return params.data_visualization;
        }
      }
    }

    return default_value;
  }

};

